mesonbuild/coredata.py

```diff
@@ -747,24 +748,39 @@ def set_default_options(self, default_options, subproject, env):
             mlog.warning('Recommend using either -Dbuildtype or -Doptimization + -Ddebug. '
                          'Using both is redundant since they override each other. '
                          'See: https://mesonbuild.com/Builtin-options.html#build-type-options')
+
         cmd_line_options = OrderedDict()
-        # Set project default_options as if they were passed to the cmdline.
+        # Set default options as if they were passed to the command line.
         # Subprojects can only define default for user options and not yielding
         # builtin option.
         from . import optinterpreter
-        for k, v in default_options.items():
+        for k, v in chain(default_options.items(), env.meson_options.host.get(subproject, {}).items()):
             if subproject:
                 if (k not in builtin_options or builtin_options[k].yielding) \
                         and optinterpreter.is_invalid_name(k, log=False):
                     continue
-                k = subproject + ':' + k
+                k = '{}:{}'.format(subproject, k)
             cmd_line_options[k] = v
 
+        # load the values for user options out of the appropriate machine file,
+        # then overload the command line
+        for project, opts in env.user_options.items():
+            for k, v in opts.items():
+                if project:
+                    k = '{}:{}'.format(project, k)
+                cmd_line_options[k] = v
+
         # Override project default_options using conf files (cross or native)
         for k, v in env.paths.host:
             if v is not None:
                 cmd_line_options[k] = v
 
+        # For cross builds we need to get the build specifc options
+        if env.meson_options.host != env.meson_options.build and subproject in env.meson_options.build:
+            for k in builtin_options_per_machine.keys():
```

Comment:
There are also compiler options that are per machine, is that handled too?

Reply:
Oh oops! I knew something was off...I of all people should have caught that since I made the relevant refactor!